src/iceberg/catalog/rest/endpoint.h

```diff
@@ -0,0 +1,144 @@
+  /// \return Endpoint instance or error if malformed.
+  static Result<Endpoint> FromString(std::string_view str);
+
+  /// \brief Get the HTTP method.
+  constexpr HttpMethod method() const { return method_; }
+
+  /// \brief Get the path template.
+  std::string_view path() const { return path_; }
+
+  /// \brief Serialize to "METHOD /path" format.
+  std::string ToString() const;
+
+  constexpr bool operator==(const Endpoint& other) const {
+    return method_ == other.method_ && path_ == other.path_;
+  }
+
+  // Namespace endpoints
+  static Endpoint ListNamespaces() {
+    return {HttpMethod::kGet, "/v1/{prefix}/namespaces"};
+  }
+  static Endpoint GetNamespaceProperties() {
+    return {HttpMethod::kGet, "/v1/{prefix}/namespaces/{namespace}"};
+  }
+  static Endpoint NamespaceExists() {
+    return {HttpMethod::kHead, "/v1/{prefix}/namespaces/{namespace}"};
+  }
+  static Endpoint CreateNamespace() {
+    return {HttpMethod::kPost, "/v1/{prefix}/namespaces"};
+  }
+  static Endpoint UpdateNamespace() {
+    return {HttpMethod::kPost, "/v1/{prefix}/namespaces/{namespace}/properties"};
+  }
+  static Endpoint DropNamespace() {
+    return {HttpMethod::kDelete, "/v1/{prefix}/namespaces/{namespace}"};
+  }
+
+  // Table endpoints
+  static Endpoint ListTables() {
+    return {HttpMethod::kGet, "/v1/{prefix}/namespaces/{namespace}/tables"};
+  }
+  static Endpoint LoadTable() {
+    return {HttpMethod::kGet, "/v1/{prefix}/namespaces/{namespace}/tables/{table}"};
+  }
+  static Endpoint TableExists() {
+    return {HttpMethod::kHead, "/v1/{prefix}/namespaces/{namespace}/tables/{table}"};
+  }
+  static Endpoint CreateTable() {
+    return {HttpMethod::kPost, "/v1/{prefix}/namespaces/{namespace}/tables"};
+  }
+  static Endpoint UpdateTable() {
+    return {HttpMethod::kPost, "/v1/{prefix}/namespaces/{namespace}/tables/{table}"};
+  }
+  static Endpoint DeleteTable() {
+    return {HttpMethod::kDelete, "/v1/{prefix}/namespaces/{namespace}/tables/{table}"};
+  }
+  static Endpoint RenameTable() {
+    return {HttpMethod::kPost, "/v1/{prefix}/tables/rename"};
+  }
+  static Endpoint RegisterTable() {
+    return {HttpMethod::kPost, "/v1/{prefix}/namespaces/{namespace}/register"};
+  }
+  static Endpoint ReportMetrics() {
+    return {HttpMethod::kPost,
+            "/v1/{prefix}/namespaces/{namespace}/tables/{table}/metrics"};
+  }
+  static Endpoint TableCredentials() {
+    return {HttpMethod::kGet,
+            "/v1/{prefix}/namespaces/{namespace}/tables/{table}/credentials"};
+  }
+
+  // Transaction endpoints
+  static Endpoint CommitTransaction() {
+    return {HttpMethod::kPost, "/v1/{prefix}/transactions/commit"};
+  }
+
+ private:
+  Endpoint(HttpMethod method, std::string_view path) : method_(method), path_(path) {}
+
+  HttpMethod method_;
+  std::string path_;
+};
+
+struct ICEBERG_REST_EXPORT EndpointHash {
+  std::size_t operator()(const Endpoint& endpoint) const noexcept {
+    return std::hash<std::string>()(endpoint.ToString());
```

Comment:
This is inefficient since we create a temp string which is unnecessary. It would be better to write it as:

```
return std::hash<int32_t>(static_cast<int32_t>(endpoint.method_)) * 31 + StringHash{}(endpoint.path_);
```